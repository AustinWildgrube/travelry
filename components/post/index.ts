export * from './Comment';
export * from './CommentList';