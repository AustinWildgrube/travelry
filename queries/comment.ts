import { supabase } from '&/services/supabase';
import { type Comment, type PaginatedData } from '&/types/types';
import { COMMENTS_PER_PAGE, REPLIES_PER_PAGE } from '&/utils/constants';

export const createComment = async (
  accountId: string,
  postId: string,
  text: string,
  reply?: boolean,
  inReplyToCommentId?: number,
  replyOriginCommentId?: number,
): Promise<Comment> => {
  const { data: comment, error } = await supabase
    .from('comment')
    .insert([
      {
        account_id: accountId,
        post_id: postId,
        text: text,
        reply: reply,
        in_reply_to_comment_id: inReplyToCommentId,
        reply_origin_comment_id: replyOriginCommentId,
      },
    ])
    .select(
      `
        id,
        text,
        post_id,
        account (
          id,
          first_name,
          last_name,
          avatar_url,
          avatar_placeholder
        ),
        comment_like (
          id,
          comment_id,
          account_id
        ),
        reply_origin_comment_id,
        in_reply_to_comment_id (
          id,
          account (
            id,
            first_name,
            last_name
          )
        )
      `,
    )
    .single();

  // There is an issue where the trigger hasn't happened yet,
  // so we need to manually add the comment_stat
  if (comment) {
    // @ts-ignore - we know it doesn't exist
    comment.comment_stat = {
      id: comment.id,
      likes_count: 0,
      replies_count: 0,
    };
  }

  if (error) {
    throw new Error(`createComment(${error.code}): ${error.message}`);
  }

  return comment as unknown as Comment;
};

export const getCommentsByPostId = async (page: number, userId: string, postId: string): Promise<PaginatedData<Comment>> => {
  const {
    data: comments,
    count,
    error,
  } = await supabase
    .from('comment')
    .select(
      `
        id,
        text,
        post_id,
        account (
          id,
          username,
          first_name,
          last_name,
          avatar_url,
          avatar_placeholder
        ),
        comment_stat (
          id,
          likes_count,
          replies_count
        ),
        comment_like (
          id,
          comment_id,
          account_id
        ),
        reply_origin_comment_id,
        in_reply_to_comment_id (
          id,
          account (
            id,
            first_name,
            last_name
          )
        )
      `,
      {
        count: 'exact',
      },
    )
    .range(0, page * COMMENTS_PER_PAGE - 1)
    .eq('comment_like.account_id', userId)
    .eq('post_id', postId)
    .eq('reply', false)
    .order('id', { ascending: false });

  if (error) {
    throw new Error(`getCommentsByPostId(${error.code}): ${error.message}`);
  }

  return {
    data: comments as Comment[],
    count: count || 0,
    cursor: page,
  };
};

export const getRepliesByCommentId = async (page: number, currentUserId: string, commentId: number): Promise<PaginatedData<Comment>> => {
  const {
    data: replies,
    count,
    error,
  } = await supabase
    .from('comment')
    .select(
      `
        id,
        text,
        post_id,
        account (
          id,
          first_name,
          last_name,
          avatar_url,
          avatar_placeholder
        ),
        comment_stat (
          id,
          likes_count,
          replies_count
        ),
        comment_like (
          id,
          comment_id,
          account_id
        ),
        reply_origin_comment_id,
        in_reply_to_comment_id (
          id,
          account (
            id,
            first_name,
            last_name
          )
        )
     `,
      {
        count: 'exact',
      },
    )
    .range(0, page * REPLIES_PER_PAGE - 1)
    .eq('comment_like.account_id', currentUserId)
    .eq('reply_origin_comment_id', commentId)
    .order('id', { ascending: true });

  if (error) {
    throw new Error(`getRepliesByCommentId(${error.code}): ${error.message}`);
  }

  return {
    data: replies as Comment[],
    count: count || 0,
    cursor: page,
  };
};

export const deleteCommentById = async (commentId: number): Promise<void> => {
  const { error } = await supabase.from('comment').delete().eq('id', commentId);

  if (error) {
    throw new Error(`${error.code}: deleteCommentById: ${error.message}`);
  }
};
