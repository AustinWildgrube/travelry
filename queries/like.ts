import { supabase } from '&/services/supabase';

export const likePost = async (accountId: string, postId: string): Promise<void> => {
  const { error } = await supabase.from('post_like').insert([
    {
      account_id: accountId,
      post_id: postId,
    },
  ]);

  if (error) {
    throw new Error(`likePost(${error.code}): ${error.message}`);
  }
};

export const unlikePost = async (accountId: string, postId: string): Promise<void> => {
  let { error } = await supabase.from('post_like').delete().eq('account_id', accountId).eq('post_id', postId);

  if (error) {
    throw new Error(`unlikePost(${error.code}): ${error.message}`);
  }
};

export const likeComment = async (accountId: string, commentId: number): Promise<void> => {
  const { error } = await supabase.from('comment_like').insert([
    {
      account_id: accountId,
      comment_id: commentId,
    },
  ]);

  if (error) {
    throw new Error(`likeComment(${error.code}): ${error.message}`);
  }
};

export const unlikeComment = async (accountId: string, commentId: number): Promise<void> => {
  let { error } = await supabase.from('comment_like').delete().eq('account_id', accountId).eq('comment_id', commentId);

  if (error) {
    throw new Error(`unlikeComment(${error.code}): ${error.message}`);
  }
};
