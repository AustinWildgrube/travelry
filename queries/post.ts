import { supabase } from '&/services/supabase';
import { type AlbumPostMedia, type PaginatedData, type Post } from '&/types/types';
import { POSTS_PER_PAGE } from '&/utils/constants';

export const getFeedPosts = async (page: number, account_id: string): Promise<PaginatedData<Post>> => {
  const { data, count, error } = await supabase
    .from('post')
    .select(
      `
        id,
        caption,
        created_at,
        location,
        account (
          id,
          username,
          first_name,
          last_name,
          avatar_url,
          avatar_placeholder
        ),
        post_media (
          id,
          file_url,
          file_placeholder
        ),
        post_stat (
          id,
          likes_count
        ),
        post_like (
           id, 
           post_id,
           account_id
        )
      `,
      {
        count: 'exact',
      },
    )
    .eq('post_like.account_id', account_id)
    .range((page - 1) * POSTS_PER_PAGE, page * POSTS_PER_PAGE - 1);

  if (error) {
    throw new Error(`getAllPosts(${error.code}): ${error.message}`);
  }

  return {
    data: data as Post[],
    count: count || 0,
    cursor: page,
  };
};

export const getPostById = async (id: string | string[] | undefined, account_id: string | undefined): Promise<Post> => {
  const { data, error } = await supabase
    .from('post')
    .select(
      `
        id,
        caption,
        created_at,
        location,
        account (
          id,
          username,
          first_name,
          last_name,
          avatar_url,
          avatar_placeholder
        ),
        post_media (
          id,
          file_url,
          file_placeholder
        ),
        post_stat (
          id,
          likes_count
        ),
        post_like (
           id, 
           post_id,
           account_id
        )
      `,
    )
    .eq('post_like.account_id', account_id)
    .eq('id', id)
    .single();

  if (error) {
    throw new Error(`getPostById(${error.code}): ${error.message}`);
  }

  return data as Post;
};

export const getPostMediaByAlbumId = async (id: string): Promise<AlbumPostMedia[]> => {
  const { data, error } = await supabase
    .from('post')
    .select(
      `
      id,
      album_id,
      post_media (
        id,
        file_url,
        file_placeholder
      )
    `,
    )
    .eq('album_id', id);

  if (error) {
    throw new Error(`getPostMediaByAlbumId(${error.code}): ${error.message}`);
  }

  return data as AlbumPostMedia[];
};

export const deletePostById = async (id: string): Promise<void> => {
  const { error } = await supabase.from('post').delete().eq('id', id);

  if (error) {
    throw new Error(`deletePostById(${error.code}): ${error.message}`);
  }
};
