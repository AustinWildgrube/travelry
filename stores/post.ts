import { create } from 'zustand';

import { type Post } from '&/types/types';

type PostState = {
  viewedPost: Post | null;
  setViewedPost: (post: Post) => void;
};

export const usePostStore = create<PostState>(set => ({
  viewedPost: null,
  setViewedPost: (viewedPost: Post) => {
    set({ viewedPost });
  },
}));
