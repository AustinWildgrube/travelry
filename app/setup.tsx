import { Fragment, useState } from 'react';

import { Link } from 'expo-router';
import { SafeAreaView } from 'react-native-safe-area-context';

import { Confetti, SetupProfileDetails, SetupProfileImage } from '&/components/auth';
import { Button, Heading, Text } from '&/components/core';
import { CONFETTI_FALL_SPEED, CONFETTI_ITEM_HEIGHT, CONFETTI_ITEM_WIDTH } from '&/utils/constants';
import { makeStyles } from '&/utils/makeStyles';

export default function Setup(): JSX.Element {
  const styles = useStyles();
  const [step, setStep] = useState(0);

  return (
    <SafeAreaView style={styles.setup}>
      <Heading>Setup your profile</Heading>

      {step === 0 ? (
        <Fragment>
          <Text style={styles.setupSlogan}>Upload an image - it's worth a thousand words</Text>
          <SetupProfileImage setStep={setStep} />

          <Button onPress={() => setStep(1)} variant="link" style={styles.skipLink}>
            <Button.Text>Skip</Button.Text>
          </Button>

          <Confetti
            fallSpeed={CONFETTI_FALL_SPEED}
            flipSpeed={3}
            horizSpeed={50}
            itemColors={['#84a98c', '#ffb703', '#023047']}
            itemDimensions={{ height: CONFETTI_ITEM_HEIGHT, width: CONFETTI_ITEM_WIDTH }}
            itemTintStrength={0.8}
            numItems={100}
          />
        </Fragment>
      ) : (
        <Fragment>
          <Text style={styles.setupSlogan}>Tell the world a little about yourself</Text>
          <SetupProfileDetails />

          <Link href="/feed" asChild>
            <Button variant="link" style={styles.skipLink}>
              <Button.Text>Skip</Button.Text>
            </Button>
          </Link>
        </Fragment>
      )}
    </SafeAreaView>
  );
}

const useStyles = makeStyles(theme => ({
  setup: {
    flex: 1,
    padding: theme.space['4'],
  },
  setupSlogan: {
    marginTop: theme.space['1'],
    marginBottom: theme.space['6'],
  },
  skipLink: {
    marginTop: theme.space['3'],
    zIndex: -1,
  },
}));
