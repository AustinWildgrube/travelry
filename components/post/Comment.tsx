import { Fragment, memo, useState, type Dispatch } from 'react';
import { TouchableOpacity, View } from 'react-native';

import { useActionSheet } from '@expo/react-native-action-sheet';
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { setStringAsync } from 'expo-clipboard';
import { impactAsync, ImpactFeedbackStyle } from 'expo-haptics';
import { Link } from 'expo-router';
import { ChevronRightIcon } from 'lucide-react-native';

import { Avatar, Button, Heading, HStack, Icon, Pressable, Text, VStack } from '&/components/core';
import { LoadingIndicator, MemoizedLikeButton } from '&/components/shared';
import { useAvatarQuery } from '&/hooks/useAvatarQuery';
import { useDoubleTap } from '&/hooks/useDoubleTap';
import { deleteCommentById, getRepliesByCommentId } from '&/queries/comment';
import { likeComment, unlikeComment } from '&/queries/like';
import { useUserStore } from '&/stores/user';
import { type Comment, type InfiniteQuery, type PaginatedData } from '&/types/types';
import { makeStyles } from '&/utils/makeStyles';

type CommentProps = {
  comment: Comment;
  setInReplyTo: Dispatch<{ id: number; name: string; origin: number } | null>;
};

function Comment({ comment, setInReplyTo }: CommentProps): JSX.Element {
  const styles = useStyles();
  const queryClient = useQueryClient();
  const doubleTapBind = useDoubleTap(() => mutateCommentLike());

  const { showActionSheetWithOptions } = useActionSheet();
  const { data: avatar } = useAvatarQuery(comment.account.id, comment.account.avatar_url);

  const [repliesPage, setRepliesPage] = useState(1);
  const [isCommentLiked, setIsCommentLiked] = useState(Object.keys(comment.comment_like).length > 0);

  const user = useUserStore(state => state.user);
  if (!user) return <LoadingIndicator />;

  const {
    data: replyData,
    refetch: refetchReply,
    isRefetching: isRefetchingReply,
  } = useQuery({
    queryKey: ['replies', comment.id],
    queryFn: () => {
      setRepliesPage(repliesPage + 1);
      return getRepliesByCommentId(repliesPage, user.id, comment.id);
    },
    enabled: false, // only fetch when the "View x Replies" button is pressed
  });

  const { mutate: mutateCommentLike } = useMutation({
    mutationFn: () => (isCommentLiked ? unlikeComment(user.id, comment.id) : likeComment(user.id, comment.id)),
    onMutate: () => impactAsync(ImpactFeedbackStyle.Medium),
    onSuccess: () => {
      if (comment.reply_origin_comment_id) {
        queryClient.setQueryData(
          ['replies', comment.reply_origin_comment_id],
          (oldData: PaginatedData<Comment> | undefined): PaginatedData<Comment> => {
            if (!oldData) return { data: [], count: 0, cursor: 0 };

            for (const oldReply of oldData.data) {
              if (oldReply.id === comment.id) {
                oldReply.comment_stat.likes_count = isCommentLiked ? oldReply.comment_stat.likes_count - 1 : oldReply.comment_stat.likes_count + 1;
                break;
              }
            }

            return oldData;
          },
        );
      } else {
        queryClient.setQueryData(
          ['comments', comment.post_id],
          (oldData: InfiniteQuery<PaginatedData<Comment>> | undefined): InfiniteQuery<PaginatedData<Comment>> => {
            if (!oldData) return { pageParams: [], pages: [{ count: 0, cursor: 0, data: [] }] };

            for (const page of oldData.pages) {
              for (const oldComment of page.data) {
                if (oldComment.id === comment.id) {
                  oldComment.comment_stat.likes_count = isCommentLiked
                    ? oldComment.comment_stat.likes_count - 1
                    : oldComment.comment_stat.likes_count + 1;
                  break;
                }
              }
            }

            return oldData;
          },
        );
      }

      setIsCommentLiked(!isCommentLiked);
    },
  });

  const hideReplies = (): void => {
    // removeQuery wasn't clearing added comments, so we must manually do it
    queryClient.setQueryData(['replies', comment.id], []);
    setRepliesPage(1);
  };

  const setReply = (comment: Comment): void => {
    setInReplyTo({
      id: comment.id,
      name: comment.account.first_name + ' ' + comment.account.last_name,
      origin: comment.reply_origin_comment_id ? comment.reply_origin_comment_id : comment.id,
    });
  };

  const openCommentActions = (): void => {
    const options = ['Copy Comment', 'Cancel'];
    const isDeletable = comment.account.id === user?.id;
    const cancelButtonIndex = isDeletable ? 2 : 1;
    const destructiveButtonIndex = isDeletable ? 1 : undefined;

    if (isDeletable) {
      options.splice(1, 0, 'Delete Comment');
    }

    showActionSheetWithOptions({ options, cancelButtonIndex, destructiveButtonIndex }, (selectedIndex: number | undefined) => {
      switch (selectedIndex) {
        case 0:
          setStringAsync(comment.text);
          break;

        case destructiveButtonIndex:
          deleteCommentById(comment.id);
          queryClient.setQueryData(['comments', comment.post_id], (oldData: Comment[] | undefined): Comment[] | undefined => {
            return oldData?.filter(oldComment => oldComment.id !== comment.id);
          });
          break;
      }
    });
  };

  return (
    <Fragment>
      <Pressable onLongPress={openCommentActions} key={comment.id} style={styles.comment} {...doubleTapBind}>
        <Link href={`/(tabs)/${comment.account.id}`} asChild>
          <Avatar as={TouchableOpacity} style={styles.commentAvatar}>
            <Avatar.FallbackText>
              {comment.account.first_name} {comment.account.last_name}
            </Avatar.FallbackText>

            <Avatar.Image source={{ uri: avatar }} placeholder={comment.account.avatar_placeholder} />
          </Avatar>
        </Link>

        <VStack style={styles.commentContainer}>
          <HStack space="xs" style={styles.commentParticipants}>
            <Link href={`/(tabs)/${comment.account.id}`}>
              <Heading size="sm">
                {comment.account.first_name} {comment.account.last_name}
              </Heading>
            </Link>

            {comment.in_reply_to_comment_id && comment.in_reply_to_comment_id.id !== comment.reply_origin_comment_id && (
              <Fragment>
                <Icon as={ChevronRightIcon} size={16} style={styles.commentReplyIcon} />

                <Link href={`/(tabs)/${comment.in_reply_to_comment_id.account.id}`}>
                  <Heading size="sm">
                    {comment.in_reply_to_comment_id.account.first_name} {comment.in_reply_to_comment_id.account.last_name}
                  </Heading>
                </Link>
              </Fragment>
            )}
          </HStack>

          <Text size="sm">{comment.text}</Text>

          <HStack style={styles.commentActions}>
            <Button onPress={() => setReply(comment)} size="sm" variant="link" style={styles.actionButton}>
              <Button.Text style={styles.actionButtonText}>Reply</Button.Text>
            </Button>

            {comment.comment_stat.replies_count > 0 && !comment.reply_origin_comment_id && (
              <Button onPress={() => refetchReply()} size="sm" variant="link" style={styles.actionButton}>
                <Button.Text style={styles.actionButtonText}>
                  {`View ${comment.comment_stat.replies_count} ${comment.comment_stat.replies_count === 1 ? 'Reply' : 'Replies'}`}
                </Button.Text>
              </Button>
            )}
          </HStack>
        </VStack>

        <MemoizedLikeButton isLiked={isCommentLiked} likeCount={comment.comment_stat.likes_count} likeAction={mutateCommentLike} iconSize={18} />
      </Pressable>

      {/* reply list: maybe move to a separate component? */}
      {isRefetchingReply ? (
        <View style={styles.loadingWrapper}>
          <LoadingIndicator size="small" />
        </View>
      ) : (
        replyData?.data?.map((reply: Comment, index: number) => (
          <View key={reply.id} style={styles.replyList}>
            <MemoizedComment comment={reply} setInReplyTo={setInReplyTo} />

            {index === replyData.data.length - 1 && (
              <HStack style={styles.replyActions}>
                {replyData.count > replyData.data.length && (
                  <Button onPress={() => refetchReply()} size="sm" variant="link" style={styles.actionButton}>
                    <Button.Text style={styles.actionButtonText}>
                      {`View ${replyData.count - replyData?.data.length} More ${
                        replyData.count - replyData?.data.length === 1 ? 'Reply' : 'Replies'
                      }`}
                    </Button.Text>
                  </Button>
                )}

                <Button onPress={hideReplies} size="sm" variant="link" style={styles.replyHideButton}>
                  <Button.Text style={styles.actionButtonText}>Hide Replies</Button.Text>
                </Button>
              </HStack>
            )}
          </View>
        ))
      )}
    </Fragment>
  );
}

const useStyles = makeStyles(theme => ({
  comment: {
    flexDirection: 'row',
    marginBottom: theme.space['6'],
  },
  commentAvatar: {
    marginRight: theme.space['3'],
  },
  commentContainer: {
    flex: 1,
  },
  commentParticipants: {
    alignItems: 'center',
  },
  commentReplyIcon: {
    color: theme.colors['textLight900'],
  },
  commentActions: {
    alignItems: 'center',
  },
  actionButton: {
    marginRight: theme.space['2'],
    paddingHorizontal: theme.space['0'],
  },
  actionButtonText: {
    color: theme.colors['trueGray500'],
    fontWeight: theme.fontWeights['semibold'],
  },
  loadingWrapper: {
    marginBottom: theme.space['8'],
  },
  replyList: {
    paddingLeft: theme.space['8'],
  },
  replyActions: {
    marginBottom: theme.space['6'],
  },
  replyHideButton: {
    flex: 1,
    justifyContent: 'flex-end',
    paddingHorizontal: theme.space['0'],
  },
}));

export const MemoizedComment = memo(Comment);
