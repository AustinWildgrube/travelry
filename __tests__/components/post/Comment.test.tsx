import { fireEvent, screen, waitFor } from '@testing-library/react-native';

import { MemoizedComment } from '&/components/post/Comment';
import { getRepliesByCommentId } from '&/queries/comment';
import { likeComment, unlikeComment } from '&/queries/like';
import { type Comment } from '&/types/types';

import { render } from '&/jest.setup';

const mockSetInReplyTo = jest.fn();

jest.mock('&/queries/like', () => ({
  likeComment: jest.fn(),
  unlikeComment: jest.fn(),
}));

jest.mock('&/queries/comment', () => ({
  deleteCommentById: jest.fn(),
  getRepliesByCommentId: jest.fn().mockResolvedValue({
    data: [
      {
        id: 173,
        text: '123',
        post_id: '5696c81a-b2c8-4e71-bd9b-6924cb0a8388',
        reply_origin_comment_id: 172,
        comment_like: [
          {
            id: 228,
            comment_id: 173,
            account_id: 'dadee07d-8db7-4429-9432-b43c8fdbb9c9',
          },
        ],
        account: {
          id: 'dadee07d-8db7-4429-9432-b43c8fdbb9c9',
          first_name: 'Cameron',
          last_name: 'Greer',
          avatar_url: '39fe83a0ce5ec6771060b1de25ea55017b55724e2ecf3160bef0b7ec25b99e8f.jpg',
        },
        comment_stat: {
          id: 142,
          likes_count: 1,
          replies_count: 0,
        },
        in_reply_to_comment_id: {
          id: 172,
          account: {
            id: 'dadee07d-8db7-4429-9432-b43c8fdbb9c9',
            first_name: 'Cameron',
            last_name: 'Greer',
          },
        },
      },
      {
        id: 174,
        text: 'Reply',
        post_id: '5696c81a-b2c8-4e71-bd9b-6924cb0a8388',
        reply_origin_comment_id: 172,
        comment_like: [],
        account: {
          id: 'dadee07d-8db7-4429-9432-b43c8fdbb9c9',
          first_name: 'Cameron',
          last_name: 'Greer',
          avatar_url: '39fe83a0ce5ec6771060b1de25ea55017b55724e2ecf3160bef0b7ec25b99e8f.jpg',
        },
        comment_stat: {
          id: 143,
          likes_count: 0,
          replies_count: 0,
        },
        in_reply_to_comment_id: {
          id: 172,
          account: {
            id: 'dadee07d-8db7-4429-9432-b43c8fdbb9c9',
            first_name: 'Cameron',
            last_name: 'Greer',
          },
        },
      },
      {
        id: 175,
        text: 'Yeah',
        post_id: '5696c81a-b2c8-4e71-bd9b-6924cb0a8388',
        reply_origin_comment_id: 172,
        comment_like: [],
        account: {
          id: 'dadee07d-8db7-4429-9432-b43c8fdbb9c9',
          first_name: 'Cameron',
          last_name: 'Greer',
          avatar_url: '39fe83a0ce5ec6771060b1de25ea55017b55724e2ecf3160bef0b7ec25b99e8f.jpg',
        },
        comment_stat: {
          id: 144,
          likes_count: 0,
          replies_count: 0,
        },
        in_reply_to_comment_id: {
          id: 172,
          account: {
            id: 'dadee07d-8db7-4429-9432-b43c8fdbb9c9',
            first_name: 'Cameron',
            last_name: 'Greer',
          },
        },
      },
    ],
    count: 14,
    cursor: 2,
  }),
}));

describe('Comment', () => {
  beforeEach(() => {
    const comment: Comment = {
      id: 1,
      text: 'Test comment',
      post_id: '1',
      reply_origin_comment_id: null,
      in_reply_to_comment_id: {
        id: 1,
        account: {
          id: '1',
          first_name: 'Rob',
          last_name: 'Stark',
        },
      },
      account: {
        id: '1',
        username: 'jdoe4',
        first_name: 'John',
        last_name: 'Doe',
        avatar_url: 'avatar.jpg',
        avatar_placeholder: 'avatar.jpg',
      },
      comment_like: {
        id: 0,
        account_id: '',
        comment_id: 0,
      },
      comment_stat: {
        id: 0,
        likes_count: 0,
        replies_count: 1,
      },
    };

    render(<MemoizedComment comment={comment} setInReplyTo={mockSetInReplyTo} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    screen.toJSON();
    expect(screen).toMatchSnapshot();
  });

  it('should render the Comment component correctly', () => {
    expect(screen.getByText(/Test comment/)).toBeTruthy();
    expect(screen.getByText(/John Doe/i)).toBeTruthy();
    expect(screen.getByText(/Rob Stark/i)).toBeTruthy(); // in reply to
  });

  // it('should show action menu on long press', async () => {
  //   fireEvent(screen.getByText(/Test comment/), 'onLongPress');
  //
  //   expect(await screen.findByText(/Copy Comment/)).toBeTruthy();
  //   expect(await screen.findByText(/Cancel/)).toBeTruthy();
  // });

  // it('should call deleteCommentById and update the query data when deleting the comment', () => {
  //   fireEvent(screen.getByText(/Test comment/), 'onLongPress');
  //   fireEvent.press(screen.getByText(/Delete Comment/i));
  //
  //   expect(deleteCommentById).toHaveBeenCalledTimes(1);
  //   expect(deleteCommentById).toHaveBeenCalledWith(1);
  //
  //   expect(useQueryClient().setQueryData).toHaveBeenCalledTimes(1);
  //   expect(useQueryClient().setQueryData).toHaveBeenCalledWith(['comments', 1], expect.any(Function));
  // });

  it('should like and unlike the post when the like button is clicked', async () => {
    fireEvent.press(screen.getByTestId('heart'));
    await waitFor(() => expect(unlikeComment).toHaveBeenCalled());

    fireEvent.press(screen.getByTestId('heart-o'));
    await waitFor(() => expect(likeComment).toHaveBeenCalled());
  });

  it('should like the comment when double clicked', async () => {
    const comment = screen.getByText(/Test comment/);
    fireEvent.press(comment);
    fireEvent.press(comment);

    // we already start with the comment liked
    await waitFor(() => expect(unlikeComment).toHaveBeenCalled());
  });

  // it('should go to the users profile when clicking their name', () => {
  //   fireEvent.press(screen.getByText(/John Doe/i));
  //
  //   TODO: test if /profile/ is the current route
  //   expect(screen.getByText(/John Doe/i)).toBeTruthy();
  // });

  it('should call setInReplyTo when clicking the reply button', () => {
    fireEvent.press(screen.getByRole('button', { name: 'Reply' }));

    expect(mockSetInReplyTo).toHaveBeenCalledTimes(1);
    expect(mockSetInReplyTo).toHaveBeenCalledWith({
      id: 1,
      text: 'Test comment',
      origin: 1,
    });
  });

  it('should call refetchReply when clicking the "View x Replies" button', async () => {
    fireEvent.press(screen.getByText('View 1 Reply'));
    await waitFor(() => expect(getRepliesByCommentId).toHaveBeenCalled());
  });

  it('should call hideReplies when clicking the "Hide Replies" button', async () => {
    fireEvent.press(screen.getByText(/View 1 Reply/));
    fireEvent.press(await screen.findByText(/Hide Replies/i));

    expect(screen.queryByText(/Hide Replies/i)).toBeNull();
  });
});
