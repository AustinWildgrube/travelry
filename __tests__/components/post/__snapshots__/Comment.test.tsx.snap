// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment should match the snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "zIndex": 1,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flexDirection": "row",
          "marginBottom": 12,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
            "gap": 2,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            accessibilityRole="link"
            href="/profile/1"
            onPress={[Function]}
            style={
              {
                "fontFamily": "Inter_600SemiBold",
                "fontWeight": "600",
              }
            }
          >
            John
             
            Doe
          </Text>
        </View>
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontFamily": "Inter_400Regular",
              "fontSize": 14,
              "fontWeight": "400",
              "lineHeight": 20,
            }
          }
        >
          Test comment
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
              "gap": 8,
            }
          }
        >
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 2,
                },
                undefined,
                false,
              ]
            }
            testID="RNE_BUTTON_WRAPPER"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": false,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "borderColor": "#84a98c",
                    "borderRadius": 2,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                    "paddingHorizontal": 0,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#5e6977",
                      "fontFamily": "Inter_600SemiBold",
                      "fontSize": 12,
                      "fontWeight": "600",
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  Reply
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 2,
                },
                undefined,
                false,
              ]
            }
            testID="RNE_BUTTON_WRAPPER"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": false,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "borderColor": "#84a98c",
                    "borderRadius": 2,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                    "paddingHorizontal": 0,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#5e6977",
                      "fontFamily": "Inter_600SemiBold",
                      "fontSize": 12,
                      "fontWeight": "600",
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  View 1 Reply
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "transform": [
                {
                  "scale": 1,
                },
              ],
            }
          }
        >
          <View
            style={
              {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              replaceTheme={[Function]}
              testID="heart"
              updateTheme={[Function]}
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#ff190c",
                        "fontSize": 18,
                      },
                      {
                        "backgroundColor": "transparent",
                      },
                      {
                        "fontFamily": "FontAwesome",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
